### PR TITLE
Add connector metrics to EXPLAIN ANALYZE verbose output

### PR DESCRIPTION
**Pull request by rzeyde-varada**
It should allow us to get quick feedback on a specific query
(without the need to "dig" into the QueryInfo JSON).

### CODE REVIEW

File: plugin/trino-memory/src/test/java/io/trino/plugin/memory/TestMemoryConnectorTest.java [hunk 1]
```diff
@@ -164,6 +164,28 @@ public void testCustomMetricsScanOnly()
...
     }
 
+    @Test
+    public void testExplainCustomMetricsScanOnly()
+    {
+        DistributedQueryRunner runner = (DistributedQueryRunner) getQueryRunner();
+        MaterializedResult result = runner.execute(
+                getSession(),
+                "EXPLAIN ANALYZE VERBOSE SELECT partkey FROM part");
+        String explainAnalyze = (String) result.getOnlyValue();
```
Comment (electrum) [inline plugin/trino-memory/src/test/java/io/trino/plugin/memory/TestMemoryConnectorTest.java:174]: `BaseConnectorTest` extends `AbstractTestQueryFramework` which has a bunch of utility methods to simplify this. The generic way is
```java
String result = (String) computeActual("EXPLAIN ...").getOnlyValue();
```

The simpler version for this case is
```java
String result = (String) computeScalar("EXPLAIN ...");
```

There's actually a more specific `assertExplainAnalyze` method that you could use here if you wanted to convert your assertion into a regex. (we should probably refactor that method to return `AbstractStringAssert<?>` so you can do your own assertions on the result...)

Comment (rzeyde-varada) [reply plugin/trino-memory/src/test/java/io/trino/plugin/memory/TestMemoryConnectorTest.java]: Thanks, TIL about `assertExplainAnalyze` :)

**Comment by sopel39** [comment]
Could we report both `io.trino.operator.OperatorStats#connectorMetrics` and `io.trino.operator.OperatorStats#metrics`?

File: core/trino-main/src/main/java/io/trino/sql/planner/planprinter/PlanPrinter.java [hunk 2]
```diff
@@ -266,6 +272,11 @@ public static String textDistributedPlan(
                 .flatMap(Collection::stream)
                 .collect(toImmutableMap(Entry::getKey, Entry::getValue));
 
+        Map<PlanNodeId, Metrics> connectorMetrics = queryStats.getOperatorSummaries().stream()
+                .filter(stats -> stats.getOperatorType().contains("Scan")) // sometimes >1 operators having the same plan node ID
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/sql/planner/planprinter/PlanPrinter.java:276]: `connectorMetrics/metrics` should be stored within `io.trino.sql.planner.planprinter.PlanNodeStats` and should go normal plan node stats route.

Comment (rzeyde-varada) [reply core/trino-main/src/main/java/io/trino/sql/planner/planprinter/PlanPrinter.java]: Good point, thanks!

**Comment by rzeyde-varada** [comment]
Many thanks, please take a look on the updated PR :)

File: core/trino-main/src/main/java/io/trino/sql/planner/planprinter/TextRenderer.java [hunk 3]
```diff
@@ -145,6 +151,19 @@ private String printStats(PlanRepresentation plan, NodeRepresentation node)
...
     }
 
+    private void printMetrics(StringBuilder output, String type, Metrics metrics)
+    {
+        if (!verbose) {
+            return;
+        }
+        Map<String, Metric> sortedMap = new TreeMap(metrics.getMetrics());
+        sortedMap.forEach((name, metric) -> {
+            if (metric instanceof LongCount) {
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/sql/planner/planprinter/TextRenderer.java:161]: please add support for other type of metrics too

File: core/trino-main/src/main/java/io/trino/sql/planner/planprinter/TextRenderer.java [hunk 4]
```diff
@@ -145,6 +151,19 @@ private String printStats(PlanRepresentation plan, NodeRepresentation node)
         return output.toString();
     }
 
+    private void printMetrics(StringBuilder output, String type, Metrics metrics)
+    {
+        if (!verbose) {
+            return;
+        }
+        Map<String, Metric> sortedMap = new TreeMap(metrics.getMetrics());
+        sortedMap.forEach((name, metric) -> {
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/sql/planner/planprinter/TextRenderer.java:160]: maybe format should be:
```
  connector metrics:
    'x' = yyy
    'y' = zzz
  metrics:
    'z' = vvv
```
instead of repeating `connector metrics` and `metric` in each line

**Review by sopel39 — COMMENTED** [review]
lgtm % comments

File: core/trino-main/src/main/java/io/trino/sql/planner/planprinter/TextRenderer.java (hunk 4 above)
Comment (rzeyde-varada) [reply core/trino-main/src/main/java/io/trino/sql/planner/planprinter/TextRenderer.java:160]: Sounds good, fixed.

File: core/trino-main/src/main/java/io/trino/sql/planner/planprinter/TextRenderer.java (hunk 3 above)
Comment (rzeyde-varada) [reply core/trino-main/src/main/java/io/trino/sql/planner/planprinter/TextRenderer.java]: Added support for `Histogram` but only for printing its `getTotal()` value - since I wasn't sure if/how should I print the percentiles' estimates (IIUC, `TDigest` doesn't seem to have a "natural" `toString` implementation).

**Comment by rzeyde-varada** [comment]
Many thanks for the review!
I have rebased the PR over latest `master` and force-pushed 8096c93880.

File: core/trino-main/src/main/java/io/trino/sql/planner/planprinter/TextRenderer.java [hunk 5]
```diff
@@ -145,6 +152,23 @@ private String printStats(PlanRepresentation plan, NodeRepresentation node)
...
+        if (!verbose || metrics.getMetrics().isEmpty()) {
+            return;
+        }
+        output.append(label).append("\n");
+        Map<String, Metric> sortedMap = new TreeMap(metrics.getMetrics());
+        sortedMap.forEach((name, metric) -> {
+            if (metric instanceof Count) {
+                output.append(format("  '%s' = %s\n", name, ((Count) metric).getTotal()));
+            }
+            if (metric instanceof Distribution) {
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/sql/planner/planprinter/TextRenderer.java:166]: it seems that `TDigestHistogram` is missing `toString`.

Is there a nice way to print percentiles (e.g: for 0, 10, 20, ..., 90, 100)

File: core/trino-main/src/main/java/io/trino/sql/planner/planprinter/TextRenderer.java [hunk 6]
```diff
@@ -135,6 +140,8 @@ private String printStats(PlanRepresentation plan, NodeRepresentation node)
         }
         output.append("\n");
 
+        printMetrics(output, "connector metrics:", nodeStats.connectorMetrics);
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/sql/planner/planprinter/TextRenderer.java:143]: use `getConnectorMetrics` instead (and below)

File: core/trino-main/src/main/java/io/trino/sql/planner/planprinter/PlanNodeStats.java [hunk 7]
```diff
@@ -44,6 +45,8 @@
     private final DataSize planNodeSpilledDataSize;
 
     protected final Map<String, OperatorInputStats> operatorInputStats;
+    protected final Metrics metrics;
```
[the remark this answers is not in the document]
Comment (rzeyde-varada) [reply core/trino-main/src/main/java/io/trino/sql/planner/planprinter/PlanNodeStats.java:48]: Done.

File: core/trino-main/src/main/java/io/trino/sql/planner/planprinter/TextRenderer.java [hunk 8]
```diff
@@ -145,6 +152,23 @@ private String printStats(PlanRepresentation plan, NodeRepresentation node)
...
 
+    private void printMetrics(StringBuilder output, String label, Metrics metrics)
+    {
+        if (!verbose || metrics.getMetrics().isEmpty()) {
+            return;
+        }
+        output.append(label).append("\n");
+        Map<String, Metric> sortedMap = new TreeMap(metrics.getMetrics());
+        sortedMap.forEach((name, metric) -> {
+            if (metric instanceof Count) {
```
[the remark this answers is not in the document]
Comment (rzeyde-varada) [reply core/trino-main/src/main/java/io/trino/sql/planner/planprinter/TextRenderer.java:163]: Sounds good, fixed.

File: core/trino-main/src/main/java/io/trino/sql/planner/planprinter/TextRenderer.java (hunk 5 above)
Comment (rzeyde-varada) [reply core/trino-main/src/main/java/io/trino/sql/planner/planprinter/TextRenderer.java:166]: Added in a separate commit.

File: core/trino-main/src/main/java/io/trino/sql/planner/planprinter/TextRenderer.java (hunk 6 above)
Comment (rzeyde-varada) [reply core/trino-main/src/main/java/io/trino/sql/planner/planprinter/TextRenderer.java:143]: Thanks, fixed.

File: core/trino-spi/src/main/java/io/trino/spi/metrics/Metrics.java [hunk 9]
```diff
@@ -75,6 +75,9 @@ public Accumulator add(Metrics metrics)
 
         public Metrics get()
         {
+            if (merged.isEmpty()) {
```
Comment (sopel39) [inline core/trino-spi/src/main/java/io/trino/spi/metrics/Metrics.java:78]: nit: this should be separate commit

**Review by sopel39 — APPROVED** [review]

**Comment by sopel39** [comment]
lgtm % please fix test

File: core/trino-spi/src/main/java/io/trino/spi/metrics/Metrics.java (hunk 9 above)
Comment (rzeyde-varada) [reply core/trino-spi/src/main/java/io/trino/spi/metrics/Metrics.java:78]: Sounds good -> 58af4c1092

**Comment by romanz** [comment]
Many thanks!
